### ISSUE TITLE
Get Consumer lag using LibrdkafkaHandle.OutQueueLength

### ISSUE THREAD

**Issue by alexandrutatagit** [confluentinc/confluent-kafka-dotnet#1161]
Hi Matt!

I found a way to get a consumer's topic lag by reading the internal property OutQueueLength found in internal LibrdkafkaHandle Handle property of the consumer. Is this a "safe" hack?
@mhowlett, can you please explain me why this property was not publicly exposed in the library?

Thank you, 
Alex

**Comment by alexandrutatagit** [comment]
I noticed that, even if this code runs on Windows, it does not run on Linux right - on Linux it will show just the partition lag

**Comment by ghetze** [comment]
Also interested to get consumer's topic lag .

**Comment by edenhill** [comment]
The OutQueueLength() should not be used for consumer lag, instead register a Statistics handler and look at the consumer_lag metric (https://github.com/edenhill/librdkafka/blob/master/STATISTICS.md), or compare position() to get_watermark_offsets(cached=yes).

**Comment by alexandrutatagit** [comment]
Thank you @edenhill !
Can you please tell me how I can do this in the C# library? Should I hack the same way I did above?

**Comment by KokaChernov** [comment]
@edenhill  Yes, could you please explain how to do so in C#?

**Comment by wxlevel** [comment]
> The OutQueueLength() should not be used for consumer lag, instead register a Statistics handler and look at the consumer_lag metric (https://github.com/edenhill/librdkafka/blob/master/STATISTICS.md), or compare position() to get_watermark_offsets(cached=yes).

How can I get message lags in by confluent-kafka-dotnet?
Below is what I estimate the number of messages by calculating highWatermark - lowWatermark. 
```csharp

  var bootstrapServers = new string[]{"10.xx.xx.xx:9092","10.xx.xx.xx:9092","10.xx.xx.xx:9092"};
  var config = new ConsumerConfig()
  {
    BootstrapServers = string.Join( ",", bootstrapServers ),
    GroupId = "xxxxx ",
    EnableAutoCommit = false,
  }; 
  var consumer = new ConsumerBuilder<Ignore, byte[]>( config ).Build();
  
  long sumLag = 0;
  foreach (var partation in consumer.Assignment)
  {
    var watermark = consumer.GetWatermarkOffsets( partation );
    sumLag += (watermark.High.Value - watermark.Low.Value);
  }

```
but I can't get consumer.Assignment result...

**Comment by mhowlett** [comment]
yes high-low watermark is consumer lag. `GetWatermarkOffsets` uses cached values, if these aren't populated as you expect, use `QueryWatermarkOffsets` which will send a query to the broker. if you aren't assigned to any partitions, you should just be able to specify the partitions of interest, IIRC.

**Comment by dogac00** [comment]
@wxlevel You have to `Subscribe` to your topic for consumer to be assigned to partitions.
